src/jit/gentree.cpp

```diff
@@ -11833,6 +11860,41 @@ void Compiler::gtGetArgMsg(
         }
         else
         {
+#ifdef _TARGET_ARM_
+            if (curArgTabEntry->isSplit)
+            {
+                regNumber argReg = curArgTabEntry->regNum;
+                if (listCount == -1)
+                {
+                    if (curArgTabEntry->numRegs == 1)
+                    {
+                        sprintf_s(bufp, bufLength, "arg%d %s out+%02x%c", argNum, compRegVarName(argReg),
+                                  (curArgTabEntry->slotNum) * TARGET_POINTER_SIZE, 0);
+                    }
+                    else
+                    {
+                        regNumber otherRegNum = (regNumber)((unsigned)(argReg) + curArgTabEntry->numRegs - 1);
+                        char      seperator   = (curArgTabEntry->numRegs == 2) ? ',' : '-';
+                        sprintf_s(bufp, bufLength, "arg%d %s%c%s out+%02x%c", argNum, compRegVarName(argReg), seperator,
+                                  compRegVarName(otherRegNum), (curArgTabEntry->slotNum) * TARGET_POINTER_SIZE, 0);
+                    }
+                }
+                else
+                {
+                    regNumber curReg = (regNumber)((unsigned)(argReg) + listCount);
+                    if (((unsigned)(curReg)) <= MAX_ARG_REG_COUNT)
+                    {
```

Comment:
Here's an alternative:
```
unsigned regArgNum = genMapIntRegNumToRegArgNum(argReg) + listCount;
if (regArgNum < MAX_REG_ARG)
{
    regNumber curReg = genMapIntRegArgNumToRegNum(regArgNum);
    sprintf_s(bufp, bufLength, "arg%d m%d %s%c", argNum, listCount, compRegVarName(curReg), 0);
}
else
{
    unsigned stackSlot = regArgNum - MAX_REG_ARG;
    sprintf_s(bufp, bufLength, "arg%d m%d out+%s%c", argNum, listCount, stackSlot, 0);
}
```

Notes:
1. `MAX_REG_ARG` is the maximum number of register arguments (for arm, that's r0, r1, r2, r3); `MAX_ARG_REG_COUNT` is the maximum number of argument registers that can be occupied by a single call argument.
2. The code above doesn't require casting between indices or offsets and registers.
3. I think the code before had a bug, by using `<=` instead of `<` -- it would allow curReg of 4, which would be r4, but the max argument register is r3.


---
In reply to: [122913222](https://github.com/dotnet/coreclr/pull/12050#discussion_r122913222) [](ancestors = 122913222)